core/include/vecmem/memory/allocator.hpp

```diff
@@ -0,0 +1,50 @@
+#pragma once
+
+#include <memory_resource>
+
+#include "vecmem/memory/resources/base_resource.hpp"
+
+namespace vecmem::memory {
+    template<typename T>
+    class polymorphic_allocator {
+    public:
+        typedef T value_type;
+
+        typedef std::false_type propagate_on_container_move_assignment;
+        typedef std::false_type is_always_equal;
+
+        polymorphic_allocator(
+            resources::base_resource * res
+        ) : alloc(res) {};
+
+        T * allocate(std::size_t size) {
+            return alloc.allocate(size);
+        };
+
+        void deallocate(T * ptr, std::size_t size) {
+            alloc.deallocate(ptr, size);
+        };
+
+        template<typename U, typename... Args>
+        void construct(U* ptr, Args&&... args) {
+            if (is_host_accessible()) {
+                alloc.construct(ptr, std::forward<Args>(args)...);
+            }
+        };
```

Comment:
We discussed this quickly last week, but amongst all the things it's inevitable to forget about a few.

I don't think we need to bother with non-host-accessible memory types in our allocators. As we discussed at that point, using an STL allocator on top of a memory block that's not host-accessible is just not making a lot of sense. (Now that I played with it for a bit.) So I think we can just use [std::pmr::polymorphic_allocator](https://en.cppreference.com/w/cpp/memory/polymorphic_allocator) as is...

Reply:
That would definitely make life a lot easier, as you say we could just use the standard `std::pmr::polymorphic_allocator`.

Reply:
Agreed. In that case we can simply use `std::pmr::polymorphic_allocator` directly. However, I am struggling to see how the CUDA example code would then look like.

Reply:
You should just remove all of the `std::vector` tests with non-host-accessible memory. Those are really useless tests...

The unit test should just make sure that the memory resources can provide non-null pointers for device-only memory blocks.